mirrord-layer/src/sockets.rs

```diff
@@ -1,240 +1,506 @@
+            if ret != 0 {
+                error!(
+                    "listen: failed to listen ret: {:?}, addr: {:?}, sockfd: {:?}",
+                    ret, result_addr, sockfd
+                );
+                return ret;
             }
-
-            write!(socket.write_socket, "a").unwrap(); // Need to write one byte per incoming connection, hence "a"
-            connections.insert(socket.read_fd, socket.address.port(), socket);
-        } else {
-            error!("No socket found for port: {}", port);
+            let sender = unsafe { HOOK_SENDER.as_ref().unwrap() };
+            match sender.blocking_send(HookMessage::Listen(Listen {
+                fake_port: result_addr.port(),
+                real_port,
+                ipv6: result_addr.is_ipv6(),
+            })) {
+                Ok(_) => {}
+                Err(e) => {
+                    error!("listen: failed to send listen message: {:?}", e);
+                    return libc::EFAULT;
+                }
+            };
+        }
+        _ => {
+            error!(
+                "listen: socket is not bound or already listening, state: {:?}",
+                socket.state
+            );
+            return libc::EINVAL;
         }
     }
+    debug!("listen: success");
+    let mut sockets = SOCKETS.lock().unwrap();
+    sockets.insert(socket);
+    0
+}
 
-    pub fn create_data_socket(&self, connection_id: ConnectionId, address: SocketAddr) -> SockFd {
-        let (read_socket, mut write_socket) = socketpair_stream().unwrap();
-        let read_fd = read_socket.as_raw_fd();
-        debug!(
-            "Accepted connection from read_fd:{:?}, write_sock:{:?}",
-            read_fd, write_socket
-        );
+unsafe extern "C" fn listen_detour(sockfd: RawFd, backlog: c_int) -> c_int {
+    listen(sockfd, backlog)
+}
+
+fn connect(sockfd: RawFd, address: *const sockaddr, len: socklen_t) -> c_int {
+    debug!("connect called");
 
-        if let Some(data) = self.pending_data.lock().unwrap().remove(&connection_id) {
-            debug!("writing pending data for connection_id: {}", connection_id);
-            write_socket.write_all(&data).unwrap();
+    let socket = {
+        let mut sockets = SOCKETS.lock().unwrap();
+        match sockets.take(&sockfd) {
+            Some(socket) => socket,
+            None => {
+                debug!("connect: no socket found for fd: {}", &sockfd);
+                return unsafe { libc::connect(sockfd, address, len) };
+            }
         }
-        let read_fd = read_socket.as_raw_fd();
-        let data_socket = DataSocket {
-            connection_id,
-            read_socket,
-            write_socket,
-            address,
-        };
-        self.data
-            .lock()
-            .unwrap()
-            .insert(read_fd, connection_id, data_socket);
-        read_fd
+    };
+
+    // We don't handle this socket, so restore state if there was any. (delay execute bind)
```

Comment:
What do you mean here with "we don't handle this socket"? Because it's for outgoing communication?

Reply:
yes